Replace Flex+Bison based zone parser with simdzone

k0ekk0ek opened the pull request:
There's still some warnings to fix and of course more tests to run. I only did very minor testing yet for NSD integration, but it's a good time to start the review as it's quite a bit (I think). This PR remove the Bison+Flex based zone parser from NSD and replaces it with [simdzone](https://github.com/NLnetLabs/simdzone).

Please note that both these changes and the current `main` branch of simdzone require review. This is also a good time to figure out how to best manage sub modules in the NSD repository. I now just did it commit based on the latest commit in master, we probably want to change that?

wcawijngaards commented on ixfr.c:
```diff
@@ -2419,45 +2362,41 @@ static void clear_temp_table_of_rr(struct domain_table* temptable,
 /* read ixfr data new SOA */
 static int ixfr_data_readnewsoa(struct ixfr_data* data, struct zone* zone,
-	FILE* in, const char* ixfrfile, struct region* tempregion,
+	struct rr *rr, zone_parser_t *parser, struct region* tempregion,
```
The `zone_parser_t`, the _t is owned by a standard, and not actually for application usage. We previously changed all of the type declarations in NSD for that reason, after that was reported. Perhaps 'zone_parser_type' or 'zone_parser' is a nice identifier.

k0ekk0ek replied:
In NSD it's a problem because no prefixing is done. simdzone uses the prefix `zone_` consistently for all exported symbols. So, while `_t` is technically reserved by the specification, I'd say chances of name conflicts are very very minimal, especially because it's common practice to use `_t`. The case is very different from using names starting with `_` or `__`.

wcawijngaards commented on ixfr.c:
```diff
@@ -2512,76 +2447,163 @@ static int ixfr_data_readoldsoa(struct ixfr_data* data, struct zone* zone,
 	}
 	ixfr_trim_capacity(&data->del, &data->del_len, &capacity);
-	return 1;
+	return 2;
```
The routine now returns 1 for type SOA. And 2 otherwise, but later on line 2561 it increases the soa_rr_count when '2' is returned. But that would increase the soa_rr_count for nonSOA types? Is that a mistake, should it be `!=` comparisons like in line 2484?

k0ekk0ek replied:
Indeed. Thanks :+1:

wcawijngaards commented on ixfr.c:
```diff
@@ -2512,76 +2447,163 @@ static int ixfr_data_readoldsoa(struct ixfr_data* data, struct zone* zone,
+	/* check SOA and also serial, because there could be other
+	 * add and del sections from older versions collated, we can
+	 * see this del section end when it has the serial */
+	if(rr->type == TYPE_SOA && soa_rr_get_serial(rr) == data->newserial) {
```
In case `!=` is desired here, this is the suggestion
```suggestion
	if(rr->type != TYPE_SOA && soa_rr_get_serial(rr) != data->newserial) {
```

k0ekk0ek replied:
Fixed.

wcawijngaards commented on ixfr.c:
```diff
@@ -2669,27 +2689,9 @@ static int ixfr_read_one_more_file(struct nsd* nsd, struct zone* zone,
 	make_ixfr_name(ixfrfile, sizeof(ixfrfile), zfile, file_num);
-	in = fopen(ixfrfile, "r");
-	if(!in) {
-		if(errno == ENOENT) {
```
The test for end of ixfr number sequence is removed, but instead there is the zone_parse call. Does that log errors, because errors have to be logged somehow? Is the code path to deal with the file after the last existing ixfr file okay, and does it then terminate without overly logging errors for that. Perhaps fopen and close the file to test for ENOENT en also the warn_if_directory call, that is there for usability for operator configuration. The return value of zone_parse is otherwise correct, the 0 return value would make things work, this is a note about logs and checks.

k0ekk0ek replied:
Yes, the parser logs errors. The idea is that the application registers a log callback function of type `zone_log_t` in `options.log.callback` that gets called if an error occurs. The function can then format/decorate the error as desired and pass it to it's actual log function. i.e. `log_msg` in the case of NSD. While in the callback, the application can use `zone_log` (or `zone_error`, `zone_warning` or `zone_info`) to log a message using the callback registered by the application. That's how it's supposed to work at least. In this case the parser should print ab error message and return `ZONE_READ_ERROR` (or `ZONE_OUT_OF_MEMORY` in case of `ENOMEM`). I did forget about implementing the printing a message part though :sweat_smile: (so, I'll fix this in an upcoming commit)

wcawijngaards replied:
But the test for ENOENT is necessary for the routine that calls the read of ixfr files. Hence that may need more attention and not just deleting the check, as it goes from a check on ENOENT to a version of the code without a check for ENOENT, and I think that likely means the case may not be handled as it should. I could be nice to handle it by itself anyway for the reason that it is part of the read of ixfr files routine.

k0ekk0ek replied:
This is fixed by NLnetLabs/simdzone#185.

wcawijngaards replied:
Apart from the parser making an error callback, the nsd side of the code has to actually handle the file nonexisting case. Because it has to detect that the ixfr file does not exist, at this place or somewhere else in the NSD code that deals with the ixfr files.

k0ekk0ek replied:
Can you elaborate? I think I'm overlooking a case(?) `ixfr_read_from_file` repeatedly calls `ixfr_read_one_more_file` until there's no more files. Judging by the code that was previously there, no distinction is made between a file not being there or an error condition(?) Therefore, the operation stays the same? At least, that was the intent. The zone parser reads the file, if it fails, `0` (zero) is returned, otherwise `1`.

wcawijngaards replied:
There is a replacement of the old lines 2677-2681 with what is now 2664, that calls 2598, that is really like old lines 2682-2683. In references:
https://github.com/NLnetLabs/nsd/pull/278/files#diff-c2dbc9ea381942e85b224edd25e15642084afe38b89872a7e14767760c399476#L2677-L2681
turns into
https://github.com/NLnetLabs/nsd/pull/278/files#diff-c2dbc9ea381942e85b224edd25e15642084afe38b89872a7e14767760c399476R2664
and
https://github.com/NLnetLabs/nsd/pull/278/files#diff-c2dbc9ea381942e85b224edd25e15642084afe38b89872a7e14767760c399476R2598
but that is like
https://github.com/NLnetLabs/nsd/pull/278/files#diff-c2dbc9ea381942e85b224edd25e15642084afe38b89872a7e14767760c399476L2682-L2683

wcawijngaards replied:
The code links did not look at pretty as I wanted. So the error code that is returned is fine, and now the parse is fixed it always prints an error. In the old sequence of code, it would not print an error, because it is not an error to find the final ixfr file, and then stop processing the list of ixfr files. Does the new code now print an error every time a list of ixfr files is read?

k0ekk0ek replied:
Thanks, I think I get what you're saying. If the file that is passed to `zone_parse` cannot be found, it is considered an error by the parser. In this case, that should not happen because it's expected behavior. To fix that, I can check if the `file` is `NULL` in `ixfr_data_log` and then simply not print the message. `ixfr_data_read` would then return `0`, which would lead to `ixfr_read_from_file` stopping the loop. I think, with the last addition of not printing the message, the behavior will be the same as before?

wcawijngaards replied:
Yes, I think that would work! The code could also call stat() on the file before opening it, is that is more convenient.

k0ekk0ek replied:
I've chosen to add a call to `stat`. Thanks :+1:

wcawijngaards commented on zonec.c:
```diff
@@ -2190,183 +150,256 @@ check_dname(zone_type* zone)
+	rdata_count = rdata_wireformat_to_rdata_atoms(
+		state->database->region, state->domains, type, rdlength, &buffer, &rdatas);
+	// rdlength must not exceed maximum of 65535 bytes
+	assert(rdlength < MAX_RDLENGTH);
```
Is there a reason why 65535 itself is not allowed? Perhaps use <= for the comparison, but that is then always true due to the integer type.

k0ekk0ek replied:
Removed this check. Thanks for pointing out the wrong assumption :+1:

wcawijngaards commented on zonec.c:
```diff
@@ -2190,183 +150,256 @@ check_dname(zone_type* zone)
+	assert(rdlength < MAX_RDLENGTH);
+	// number of atoms must not exceed maximum of 65535 (all empty strings)
+	assert(rdata_count >= 0);
+	assert(rdata_count < MAX_RDLENGTH);
```
Is there a reason why 65535 itself is not allowed? Perhaps use <= for the comparison, but that is then always true due to the integer type.

k0ekk0ek replied:
Actually, `rdata_count` is an `ssize_t` (unlike comment above), so the limit is higher. Thanks for spotting my wrong assumption!

wcawijngaards replied:
The comment is on the 'rdlength' line. But nice that the also included line on rdata_count can be improved!

wcawijngaards replied:
But that is another comment, for a similar issue.

wcawijngaards commented on zonec.c:
```diff
@@ -2190,183 +150,256 @@ check_dname(zone_type* zone)
+	/* we have the zone already */
+	if (type == TYPE_SOA) {
+		if (domain != state->zone->apex)
+			zone_log(parser, priority, "SOA record with invalid domain name, '%s' is not '%s'", "x", "y");
```
Placeholder "x" and "y" are here instead of the strings.

k0ekk0ek replied:
Good catch

wcawijngaards replied:
The fix prints `state->zone->apex` twice, but not `domain`.

k0ekk0ek replied:
Fixed

wcawijngaards commented on zonec.c:
```diff
@@ -2190,183 +150,256 @@ check_dname(zone_type* zone)
+	const struct dname *dname;
+	struct domain *domain;
+	struct buffer buffer;
+	int priority = ZONE_ERROR << (parser->options.non_strict == true);
```
So this compares the boolean with a 'true' value, uses the numerical value of the comparison, then uses that shifts define the ZONE_ERROR and ZONE_WARNING priorities in zone.h. It could also be with '?', or is that to stop alternate code paths, for optimization? In which case that could be annotated; that non_strict has to have 'true', not unzero, and the shifts are used in this manner. Maybe use '1' instead of 'true'?

k0ekk0ek replied:
Yeah, it's probably better to change this for readability. Will fix.

wcawijngaards commented on zonec.c:
```diff
@@ -2190,183 +150,256 @@ check_dname(zone_type* zone)
+	} else if (!zone->soa_rrset || zone->soa_rrset->rr_count == 0) {
+		log_msg(LOG_ERR, "zone configured as '%s' has no SOA record", name);
+		state.errors++;
+#if 0
```
This check is not disabled in the original zonec_read function.

k0ekk0ek replied:
Correct. I disabled it for now. I'll have to turn the owner into a `dname` and then compare. I'll fix in an upcoming commit.

k0ekk0ek replied:
I've re-enabled this check btw. Comparing the owner to the origin, which is constructed from the zone name.

wcawijngaards commented on zonec.c:
```diff
@@ -2190,183 +150,256 @@ check_dname(zone_type* zone)
-	parser_flush();
-	return errors;
-}
+	startzonec = time(NULL);
```
This time seems to be unused.

k0ekk0ek replied:
Removed

wcawijngaards reviewed (approved):
The code looks nice! There are some small notes. The submodule was causing ci to fail, also makedist.sh needs to include submodules in the source tarball output.

k0ekk0ek commented:
Rebased on current `master` and fixed `soa_test.tdir/soa_test.zone5` (no previous owner). Only test failing is `bug137_compile_outdir.tdir` which has to do with compilation, not actual functionality.

wcawijngaards commented on zonec.c:
```diff
@@ -265,7 +265,7 @@ int32_t zonec_accept(
 		if (type != TYPE_RRSIG && ttl != rrset->rrs[0].ttl) {
 			zone_log(parser, priority, "%s TTL %"PRIu16" does not match TTL %u of %s RRset",
-				domain_to_string(domain), ttl, rrset->rrs[0].ttl,
+				domain_to_string(domain), (uint16_t)ttl, rrset->rrs[0].ttl,
```
The ttl variable is uint32_t, not uint16_t.

k0ekk0ek replied:
Indeed. Format string was incorrect. Fixed.

k0ekk0ek commented:
All checks pass. I'll rebase and squash the debug commits so the PR can be merged.

Commit:
Replace Flex+Bison based parser with simdzone

k0ekk0ek commented:
I've consolidated all commits. Tests are all passing, review comments have all been addressed. I think this can be merged.
